pycaret/internal/pipeline.py

```diff
@@ -87,6 +88,53 @@ def __getattr__(self, name: str):
         # override getattr to allow grabbing of final estimator attrs
         return getattr(self._final_estimator, name)
 
+    def __getstate__(self):
+        try:
+            state = super().__getstate__()
+            state.update(self.__dict__)
+        except AttributeError:
+            state = self.__dict__.copy()
+
+        return dict(state.items(), _pycaret_versions=self._pycaret_versions)
+
+    def __setstate__(self, state):
+        pickle_versions = state.get("_pycaret_versions", {})
+        if pickle_versions != self._pycaret_versions:
+            import warnings
+
+            warnings.warn(
+                "Version mismatch:\ncurrent: {}\npickle: {}".format(
+                    self._pycaret_versions, pickle_versions
+                )
+            )
+        try:
+            super().__setstate__(state)
+        except AttributeError:
+            pass
+
+        self.__dict__.update(state)
+
+    @property
+    def _pycaret_versions(self):
+        from importlib import import_module
+
+        versions = {"pycaret": __version__}
+        ml_modules = [
```

Comment:
What about the models in `requirements-optional.txt`, like xgboost and catboost? How would we handle those?
@ngupta23 Would this check be required for time-series models as well?

Reply:
@tvdboom @ngupta23 Thank you for the review!
Please let me know if there are any other libraries you would like to add.

Reply:
I have an idea to make this work quicker and cover all the libraries. Will make a PR to lay down the groundwork.

Reply:
https://github.com/pycaret/pycaret/pull/2782 - with this we can just use `_get_deps_info` and `_get_sys_info` from `pycaret/utils/_show_versions.py` to quickly and easily obtain consistent environment information we can put into the pickle file.

Reply:
@rhoboro let me know what you think. The changes in this PR can be made right now as the API for those two functions is unchanged.

Reply:
pyod.version

@Yard1 
Thank you for the improvement.
I checked _show_versions.py during the first implementation, however it did not meet our purpose, so I listed the important libraries individually.

Is this what you mean?
Note that pyod has pyod.version.__version__ but not prod.__version__.

Please let me know if there are any areas that need further improvement.

```python
from pycaret.utils._show_versions import _get_deps_info, _get_sys_info
...
    @property
    def _pycaret_versions(self):
        return {
            "pycaret": __version__,
            "deps_info": _get_deps_info(optional=True),
            "sys_info": _get_sys_info()
        }
```

Reply:
Yes, I think this should be fine. You can convert the string versions to LooseVersion for comparison purposes

Reply:
@Yard1 
I tried with #2728 and got this result.

```python
# use _get_deps_info(optional=False) and _get_sys_info()
>>> pprint.pprint(loaded_model._pycaret_versions)
{'deps_info': {'IPython': '8.4.0',
               'category_encoders': '2.5.0',
               'imblearn': '0.9.1',
               'ipywidgets': '7.7.1',
               'jinja2': '3.1.2',
               'joblib': '1.1.0',
               'kaleido': '0.2.1',
               'lightgbm': '3.3.2',
               'matplotlib': '3.5.2',
               'numba': '0.55.2',
               'numpy': '1.21.6',
               'pandas': '1.4.3',
               'pip': '21.1.1',
               'plotly': '5.9.0',
               'pmdarima': '1.8.5',
               'psutil': '5.9.1',
               'pycaret': '3.0.0rc3',
               'pyod': '1.0.2',
               'requests': '2.28.1',
               'scikitplot': '0.3.7',
               'scipy': '1.8.1',
               'setuptools': '56.0.0',
               'sklearn': '1.1.1',
               'sktime': '0.11.4',
               'statsmodels': '0.13.2',
               'tbats': '1.1.0',
               'tqdm': '4.64.0',
               'yellowbrick': '1.4'},
 'pycaret': '3.0.0.rc3',
 'sys_info': {'executable': '/Users/rhoboro/go/src/github.com/rhoboro/pycaret/venv/bin/python3',
              'machine': 'macOS-12.4-arm64-arm-64bit',
              'python': '3.8.10 (v3.8.10:3d8993a744, May  3 2021, 09:09:08)  '
                        '[Clang 12.0.5 (clang-1205.0.22.9)]'}}
```

- deps_info: Ideal. Thank you very much.
- pycaret: I'm going to delete it because deps_info includes it.
- sys_info: This gives too many details. So, I suggest a change to `'python' : {'version': python_version(),'machine': platform.machine()}` instead of sys_info.

Finally, It will look like this.
What do you think about this?
If all is well, I will push the changes.

```
{'deps_info': {'IPython': '8.4.0',
               'category_encoders': '2.5.0',
               'imblearn': '0.9.1',
               'ipywidgets': '7.7.1',
               'jinja2': '3.1.2',
               'joblib': '1.1.0',
               'kaleido': '0.2.1',
               'lightgbm': '3.3.2',
               'matplotlib': '3.5.2',
               'numba': '0.55.2',
               'numpy': '1.21.6',
               'pandas': '1.4.3',
               'pip': '21.1.1',
               'plotly': '5.9.0',
               'pmdarima': '1.8.5',
               'psutil': '5.9.1',
               'pycaret': '3.0.0rc3',
               'pyod': '1.0.2',
               'requests': '2.28.1',
               'scikitplot': '0.3.7',
               'scipy': '1.8.1',
               'setuptools': '56.0.0',
               'sklearn': '1.1.1',
               'sktime': '0.11.4',
               'statsmodels': '0.13.2',
               'tbats': '1.1.0',
               'tqdm': '4.64.0',
               'yellowbrick': '1.4'},
 'python':  {'version': '3.8.10',
             'machine': 'arm64'}
```

Reply:
This looks good! I'd not check for machine or python version, though, that shouldn't impact anything. It's fine to keep that info in though

Reply:
Updated!!

Reply:
@Yard1 What is the status of this PR? If you need anything else, let me know and I will do it :)

Reply:
Can we only check for `deps_info`?

Reply:
@Yard1 Of course. I have pushed it, so please check it out.